lib/phoenix_library/books.ex

```diff
@@ -0,0 +1,202 @@
+
+    - params: Map com parametros a serem atualizados no livro. Deve conter uma chave
+      "id" com o id do livro a ser atualizado.
+
+  ## Exemplos
+
+    iex> PhoenixLibrary.Books.update_book(%{
+           "id" => "0200e8de-69c0-4d88-982c-31721cb4af5c",
+           "title" => "Adopting Elixir - From Concept to Production"
+         })
+
+    {:ok, %PhoenixLibrary.Books.Book{...}}
+
+    iex> PhoenixLibrary.Books.update_book(%{
+           "id" => "521675b8-df5d-495d-b579-4f2c69716318",
+           "title" => "Adopting Elixir - From Concept to Production"
+         })
+
+    {:error, :not_found}
+  """
+  def update_book(%{"id" => id} = params) do
+    case load_book(id) do
+      {:error, :not_found} -> {:error, :not_found}
+      {:ok, entity} -> update_entity(entity, params)
+    end
+  end
+
+  @doc """
+  Retorna uma tupla {:ok, <livro>} caso o livro seja encontrado no banco de dados
+  ou uma tupla {:erro, :not_found} caso o livro não possa ser encontrado.
+
+  ## Parâmetros
+
+    - id: id do livro a ser buscado no formato UUID.
+
+  ## Exemplos
+
+    iex> PhoenixLibrary.Books.delete_book("0200e8de-69c0-4d88-982c-31721cb4af5c")
+
+    {:ok, %PhoenixLibrary.Books.Book{...}}
+
+    iex> PhoenixLibrary.Books.update_book("521675b8-df5d-495d-b579-4f2c69716318")
+
+    {:error, :not_found}
+  """
+  def delete_book(id) do
+    case load_book(id) do
+      {:error, :not_found} -> {:error, :not_found}
+      {:ok, book} -> Repo.delete(book)
+    end
+  end
+
+  @doc """
+  Retorna uma lista com todos os livros cadastrados no banco de dados.
+
+  ## Exemplos
+
+    iex> PhoenixLibrary.Books.list_books()
+
+    [%PhoenixLibrary.Books.Book{...}, %PhoenixLibrary.Books.Book{...}]
+  """
+  def list_books do
+    Book
+    |> Repo.all()
+  end
+
+  @doc """
+  Retorna uma lista com todos os livros cadastrados no banco de dados cujo
+  título corresponde ao termo buscado.
+
+  ## Parâmetros
+
+    - search_term: String que representa o termo a ser buscado.
+
+  ## Exemplos
+
+    iex> PhoenixLibrary.Books.by_title("elixir")
+
+    [%PhoenixLibrary.Books.Book{...}, %PhoenixLibrary.Books.Book{...}]
+
+    iex> PhoenixLibrary.Books.by_title("Pascal")
+
+    []
+  """
+  def by_title(search_term) do
+    wildcard_search = "%#{search_term}%"
+
+    from(book in Book, where: ilike(book.title, ^wildcard_search))
+    |> Repo.all()
+  end
+
+  @doc """
+  Retorna uma lista com todos os livros cadastrados no banco de dados cuja
+  autoria corresponde ao termo buscado.
+
+  ## Parâmetros
+
+    - search_term: String que representa o termo a ser buscado.
+
+  ## Exemplos
+
+    iex> PhoenixLibrary.Books.by_authorship("valim")
+
+    [%PhoenixLibrary.Books.Book{...}, %PhoenixLibrary.Books.Book{...}]
+
+    iex> PhoenixLibrary.Books.by_title("Cantú")
+
+    []
+  """
+  def by_authorship(search_term) do
+    wildcard_search = "%#{search_term}%"
+
+    from(book in Book,
+      where: ilike(fragment("ARRAY_TO_STRING(?, ',')", book.authorship), ^wildcard_search)
```

Comment:
Poderia me explicar a ideia de usar essa função fragment e o array_to_string? To curioso

Reply:
Posso estar errado, mas acho que é porque `authorship` é um array de strings no database. Então a query vai buscar se algum elemento desse array faz match com `%#{search_term}%`.

Reply:
Se for esse caso @odineiramone voce pode fazer where(^authorship in book.authorship)

Reply:
Ele está usando o `ilike`, com o `where` direto ele vai perder o comportamento de buscar um termo dentro da string.

Reply:
Hmm, verdade 😅

Reply:
@pgr-nascimento, não sei explicar exatamente o que essa linha faz. Eu vou tentar dar a melhor explicação que eu puder mas é bem isso que o @Cendrao disse.

O campo `authorship` é um array, essa query ai me possibilita pesquisar dentro desse array. O Ecto por sí só não tem uma forma de fazer pesquisas dentro de um campo no banco representado por um array.

Ai que entra a função [`fragment/1`](https://hexdocs.pm/ecto/Ecto.Query.API.html#fragment/1). Ela me permite mandar um pedaço de SQL diretamente para o banco, que no meu caso é o `ARRAY_TO_STRING(?, ',')` interpolado com o `book.authorship`. Essa função SQL `ARRAY_TO_STRING` combina os itens de um array, que no nosso caso é o `book.authorship`, em uma string com os itens separados por vírgula.

Ai assim eu consigo fazer pesquisas dentro desse campo de array.

E só dando uma explicação a mais aqui. Eu escolhi fazer o campo de autoria dessa forma em vez de ter uma tabela autoria porque no exercício estava proposto assim e eu  queria manter a simplicidade do exercício 😅

Reply:
Brabo! Obrigado pela explicação meu lindo <3